[screengrab] Upgrade to AndroidX

colintheshots opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
Android Gradle Plugin 3.5.0 UI tests now require AndroidX dependencies. These conflict with the Screengrab uiautomator dependency in a way that makes it impossible for them to co-exist. See #14897.

Note that the minimum API level had to increase along with the minimum for AndroidX. Users of APIs 8 to 17 will need to stay on an older version of the library.

<!-- If it fixes an open issue, please link to the issue here. -->
https://github.com/fastlane/fastlane/issues/14897

### Description
<!-- Describe your changes in detail. -->
I upgraded both the Screengrab library and the example to work with the latest Android APIs and AndroidX libraries. I updated to the latest Falcon library dependency.

<!-- Please describe in detail how you tested your changes. -->
I verified that the code builds and all tests work.

### Testing Steps
<!-- Optional: steps, commands, or code used to test your changes. -->
<!-- Providing these will reduce the time needed for testing and review by the fastlane team. -->

colintheshots committed:
For #14897: Update Screengrab to AndroidX

janpio commented:
Context: "APIs 8 to 17" means Android 3.x to 4.2

colintheshots commented:
I hate to be *that* guy, but our UI tests were broken by this issue. If we downgrade Android Gradle Plugin, we lose important performance benefits from the new R8. How likely can we get a new version of Screengrab built to fix this in the next week or so?

janpio commented:
My personal problem with AndroidX is that I don't understand the backwards compatibility story. If we decide to merge this, we basically break Screengrab for everyone testing Android <= 4.2, correct?

colintheshots commented:
I think the solution is to add a note that apps supporting screengrabs before 4.3 need to use library version `1.2.0`.

The issue isn't just a few fringe apps using AndroidX test dependencies anymore. As of the new Android Studio stable version, the current version of Screengrab will break for nearly everyone.

janpio commented:
That sounds viable.

I guess this PR then should also update https://github.com/fastlane/fastlane/blob/master/screengrab/version.properties. Would something like this be considered a breaking/major change in Android land and justify a `2.0.0` then?

Documentation for `screengrab` lives in https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/docs/capture_android_screenshots.md. The comment to use `1.x` would probably be best placed at https://docs.fastlane.tools/actions/screengrab/#gradle-dependency, and maybe an additional headline below https://docs.fastlane.tools/actions/screengrab/#improved-screenshot-capture-with-falcon?

colintheshots commented:
I think this justifies a 2.0.0.

I updated the documentation. Would it make more sense to switch the default strategy to UIAutomator or to Falcon? I just updated the docs to UI Automator, since it's Google-supported and API 18 will definitely be an option for all users of 2.0.0. However, the docs seem to indicate Falcon might be better?

janpio commented:
Thanks!

I don't know re UIAutomator/Falcon. Ping @joshdholtz, you seem to have written the docs on that.

I would be good merging with default UIAutomator.

colintheshots commented:
I think we can review the code as it stands. If default UIAutomator isn't ideal, we'll hear about it.

cboisen reviewed (changes requested):
The [documentation](https://docs.fastlane.tools/actions/screengrab/) states that the argument `test_instrumentation_runner` has a default of `android.support.test.runner.AndroidJUnitRunner`. 

That should be updated to `androidx.test.runner.AndroidJUnitRunner`, right?

colintheshots commented:
> The [documentation](https://docs.fastlane.tools/actions/screengrab/) states that the argument `test_instrumentation_runner` has a default of `android.support.test.runner.AndroidJUnitRunner`.
> 
> That should be updated to `androidx.test.runner.AndroidJUnitRunner`, right?

This was updated by the PR in the example. The PR is not landed yet, however. Until it is, you won't see that change.

colintheshots commented:
Could anyone give this a needed review please?

VitaminPSG commented:
Can some one have a look to this MR, AndroidX was add year ago, but still no possibility to use with fastlane.

cboisen reviewed (changes requested):
I think the minSdk means that there's obsolete code in a couple of places. I sent a screenshot showing the lint report I get.

Also: I have some issues getting the example project tests to compile, so I cannot verify that it still works. Probably just me not knowing what to do to get it running :man_shrugging:

hauthorn commented on screengrab/screengrab-lib/build.gradle:
```diff
@@ -18,10 +18,10 @@ apply plugin: 'com.android.library'
 
     defaultConfig {
-        minSdkVersion 8
+        minSdkVersion 18
```
@colintheshots I think it's worth spending the time to simplify the statements and remove the obsolete code as a result of it.

colintheshots replied:
Done

colintheshots commented:
@hauthorn I made the changes, pushed, and rebased.

colintheshots commented:
Okay, now the tests are also fixed. Updating Falcon seems to have fixed the reflection-based code.

cboisen reviewed (approved):
It seems like it's good to go now. Good job :+1: :clap:

colintheshots commented:
@janpio Could we now get this merged? Thanks!

colintheshots commented:
@janpio Friendly reminder. This PR just requires a merge. It has two approvals now.

janpio commented:
Yes, none of those are from team members of the fastlane team. I know this is not the answer you want to hear here, but it's the only one I can currently give you - sorry. I will try to get to this asap - but no ETA.

hauthorn commented:
@janpio That's cool. The CI doesn't really run any of the screengrab-related code, so it also meant that I/we had to guess how to verify that the changes work as intended.

colintheshots commented:
It's been about a month since the code revisions were finished. Could someone please review? We'd like to use this library for several projects.

isabelrios commented:
Hello @janpio @hauthorn, wondering if you have any updates about this PR and its testing. It has been almost one month and we are blocked on getting this library working for several projects as @colintheshots mentioned. Any idea when/if this could be merged? 

Thanks!

hauthorn commented:
@isabelrios please refer to the [comment](https://github.com/fastlane/fastlane/pull/15217#issuecomment-535095483) @janpio made above. I'm not a member of the fastlane core team, so I'm not in a position to help you.

rpappalax commented:
> Yes, none of those are from team members of the fastlane team. I know this is not the answer you want to hear here, but it's the only one I can currently give you - sorry. I will try to get to this asap (Actually test it myself) - but no ETA.

Thanks @janpio.  Do you by chance know whom on the fastlane team we might reach out to for final approval?  this is a big blocker for us and we just want to do all we can

colintheshots commented:
@joshdholtz Could you please recommend anyone to do an Android PR review here? A lot of Android projects cannot take screengrabs right now and it's been like this for months.

felipe-silvestre-morais commented:
@colintheshots I have no hope this PR will be merged soon, could you provide the steps to build it locally?

Thanks

colintheshots commented:
It looks to me like someone needs to take over maintainership or fork at this point. All Android PRs seem blocked. I'll look into it.

taquitos commented:
I'm digging into this, I haven't run screengrab in a while, I'm sorry this has taken so long.

taquitos commented:
I was able to check out, build and test this using the example project with some small changes to the `Screengrabfile`. I can't get the screenshots, which I'm assuming is because I'm not doing something correctly. 
`Screengrabfile` changes:
```
tests_apk_path 'build/outputs/apk/example-debug-androidTest.apk'
app_apk_path 'build/outputs/apk/example-debug.apk'
```
to:
```
tests_apk_path 'build/outputs/apk/androidTest/debug/example-debug-androidTest.apk'
app_apk_path 'build/outputs/apk/debug/example-debug.apk'
```
I was running `./gradlew assembleAndroidTest` and `./gradlew build`

Output:
```
[09:49:01]: Pulling captured screenshots from the device
ls: /sdcard/tools.fastlane.localetester/screengrab: No such file or directory
[09:49:01]: Exit status: 1
ls: /data/data/tools.fastlane.localetester/app_screengrab: Permission denied
[09:49:01]: Exit status: 1
[09:49:01]: Make sure you've used Screengrab.screenshot() in your tests and that your expected tests are being run.
```

joshdholtz commented:
Hello 👋

I will get this and the rest of the Android PRs that time it deserves next week 😊 A majority of my time and brainpower has gone to the Android Publisher V2 to V3 migration (https://github.com/fastlane/fastlane/pull/15269) and that is getting finished up this week 🙏 I'll move my focus over to the rest of these Android PRs after that.

Sorry for the wait and I appreciate the patience ❤️

joshdholtz reviewed (approved):
@colintheshots This is 🔥 Thanks so much for making this change and my apologies again on taking so long. I have a few more screengrab PRs to get through today and then I will pop out a new release.
